harvester/file.py

```diff
@@ -152,18 +160,30 @@ def download(self, base_path=None, file_dir=None, filename=None):
          ^^^^^^^^^^^ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
           base_path          file_dir                         filename
 
-        :param dc_identifier: Dublin Core identifier for the binding to which this file
-            belongs. These identifiers are of form
-            https://digi.kansalliskirjasto.fi/sanomalehti/binding/[BINDING ID] and thus
-            the base of the download URL.
-        :type dc_identifier: str
+        :param download_function: Download function to download the file with
+        :type download_function: function
+        :param sftp_client: SFTPClient to connect to the remote host (required for remote download)
+        :type sftp_client: paramiko.SFTPClient, optional
         :param base_path: The root directory for the file structure for
             downloaded bindings.
         :type base_path: str, optional
         :param file_dir: Output directory for the files, relative to ``base_path``
         :type file_dir: str, optional
         :param filename: Output file name.
         :type filename: str, optional
+
+        """
+        download_function(
```

Comment:
Actually now that I read paramiko docs trying to figure out the IO vs OS error thingy, it looks like we could actually just pass the file from either "normal" `open` operation or a remote file from paramiko, as both look the same when it comes to writing: https://docs.paramiko.org/en/stable/api/sftp.html#paramiko.sftp_client.SFTPClient.file. That way `File` doesn't have to worry about the destination at all, just write where instructed, and Airflow can be 100% responsible of magic involved with remote files.

Pros: the implementation is a lot simpler
Cons: we have to ditch the base path + file dir + filename logic. Though I don't know if it was *that* handy to begin with, it shouldn't be a catastrophe if the caller is responsible for ensuring that ALTOs and other files end up in reasonable locations under a shared root directory.

Sorry for hemming and hawing, I didn't come to think about remote files possibly being usable just like "normal" files are.

Reply:
I think I managed to improve this quite a bit, without ditching the base path + file dir + filename logic. There's now only one download function that takes either `open` or `SFTPClient.file` as argument. 

Let me know if the new `download` function is what you had in mind!

Reply:
Sorry, I looked at this yesterday but apparently forgot to comment!

I think it's better now, but ideally we'd get rid of the whole `sftp_client` in `Mets.download`. I think it would improve readability as we get rid of the if-else and an argument that creates a pretty big shift in what the method does. It also maintains a clear division of responsibilities as `Mets` does not take up a part of the networking responsibilities.

Ideally the download function would be pretty lighweight, something like
```
def download(self, output_file, chunk_size=1024 * 1024):
    with requests.get(self.download_url, timeout=5) as source:
        source.raise_for_status()
        for chunk in source.iter_content(chunk_size=chunk_size):
            output_file.write(chunk)
```
Perhaps with checking that output_file is writable and raising an informative exception if not, I'm not sure if that would be useful or not. Might be overkill.

This would of course leave the path construction and creating the intermediate dirs for the caller, but I don't think that would be too bad. It's not likely too difficult to extract the current base + filedir + name thingy into `utils` or something if we really need that general helper function, or just construct the path locally.

Also as an afterthought, we could maybe have the default value of `chunk_size` be `None` so the default would be more "neutral", as we don't really get benefits from manually setting the chunk size unless the chunk matches the buffering of the output file.

Reply:
I did another round of changes based on these comments. The `download` function now pretty much looks like your example. File location construction and making sure that intermediate directories exist are now done in the DAG functions before calling `download`.

Reply:
Looking good!